### ISSUE TITLE
IStartupConfigureServicesFilter is not executed when Configure is used

### ISSUE THREAD

**Issue by pakrym** [aspnet/Hosting#1242]
https://github.com/aspnet/Hosting/blob/872fc80d567892096f6cd6925600285570d866fe/test/Microsoft.AspNetCore.Hosting.Tests/StartupManagerTests.cs#L141-L166

There no reason why we can't do it except internal implementation details.

/cc @Tratcher @javiercn

**Comment by pakrym** [comment]
This feature is very limited and interface name similar to `IStartupFilter` makes it even more misleading, I propose moving it to `.Internal` namespace.

**Comment by javiercn** [comment]
This is by design.
There are three patterns you can use to configure services in your application:
* Through configure services on IWebHostBuilder.
* Through configure services when you implement IStartup.
* Through ConfigureServices and ConfigureContainer<T> on a convention based Startup.

IStartupConfigureServices and IStartupConfigureContainer are designed to work only on convention based startup.

There is no general way of applying filters to service configurations for multiple reasons:
* ConfigureServices on IWebHostBuilder puts services on the hosting container, where its to early for IStartupServicesFilters to be applied.
* ConfigureServices on IStartup returns an IServiceProvider and prevents the application of filters. One posibility on this scenario is for the user to inject IEnumerable<IConfigureStartupServicesFilters> on the constructor of their IStartup implementation and apply the filtering themselves.
* ConfigureServices on a conventional Startup class need to have the signature `void ConfigureServices(IServiceCollection services)` if it decides to build and return an IServiceProvider instance itself is impossible to apply the filters.

All these factors were taken into account at the time when we designed this feature and at that time we decided that all filters for services would work only on conventional startup based classes.

It is important to note that at the time we pondered the like-hood of users incurring in each of the 3 scenarios described above and evaluated that conventional based Startup is the most likely scenario, given that it is what we prescribe in all of our templates, and that this is a pattern that the majority of our users will not customize.

The unique value of filtering services is given by their ability to override services previously configured in Startup code by the user. This is something that no library should do without and explicit gesture from the user, and the place to do so is the webhost builder, given the current DI design in which we have two containers. 

Any other services configuration related pattern can be achieved in some other way. For example, providing default services can be achieved by adding services to the services collection before the user adds their services. Then, providing a gesture the user can do when configuring the services to override such defaults.

So to summarize:
* Works as expected in the context it was designed for.
* Covers the vast majority of our user scenarios.
* There are alternative patterns our customers can use to achieve other scenarios supported by this pattern in non conventional startup scenarios.

**Comment by javiercn** [comment]
To name a few:
* I'm building a web app that spans multiple assemblies and each assembly defines their own set of services. Then on my app assembly I have a gesture a one line gesture that I perform on IWebHostBuilder to plug in such services.
* I have a library that wants to provide a gesture to in-conditionally alter the configured services in your app after startup code has run.
* I have multiple methods to setup my WebHostBuilder based on different configurations or any other factor and I want to ensure that a set of services is used for a specific configuration independently of the Startup code (which might have defaults).
* I want to build a modular app and allow each module to define their services through a filter.
* Pretty much any other scenario in which you want a set of components to collaborate for configuring the list of services.

**Comment by pakrym** [comment]
> I have a library that wants to provide a gesture to in-conditionally alter the configured services in your app after startup code has run.

This is exactly what ApplicationInsights people tried to do in their hosting startup until they realized that it will not work in some cases and prevent app from starting in the other.

In all other cases you control the startup and are able to add things to it in any order you wan't without requiring an abstraction.
This feature works only when you control kind of startup you have and not useful for libraries that want to support all scenarios.

**Comment by muratg** [comment]
Per triage/mini-design meeting, we'll try the following:

- Move `IStartupConfigureServicesFilter` to `TestHost` and make it internal
- Remove `IStartupConfigureContainerFilter` (It was put in place for completeness)
- Add a implementation of the `IServiceProviderFactory` that decorates the registered one
- Apply this to both the app container and the host container

**Comment by muratg** [comment]
@javiercn Is this coming in in the next two days?

cc @mkArtakMSFT

**Comment by mkArtakMSFT** [comment]
Hey @muratg. I've talked to @javiercn and would like to move this to rc milestone. If you have no objections, please do so. Thanks!

**Comment by muratg** [comment]
@javiercn ping :)  Coming in the next two days?

**Comment by javiercn** [comment]
@muratg I'm looking into this

**Comment by muratg** [comment]
@javiercn this would require a shiproom approval now.

**Comment by javiercn** [comment]
@muratg Yep. I'll fill in the template

**Comment by javiercn** [comment]
For 2.1 we've moved the interfaces to the internal namespace.
Keeping this bug open so that we can revisit in the future.

**Comment by muratg** [comment]
@javiercn Are you doing this in 2.2?
